src/libraries/BytesLib.sol

```diff
@@ -0,0 +1,42 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.0;
+
+/// @title Library for Bytes Manipulation
+library BytesLib {
+    error SliceOutOfBounds();
+
+    bytes4 constant SLICE_ERROR_SELECTOR = 0x3b99b53d;
+
+    /// @notice Performs the equivalent of `abi.decode(data, (uint256[], bytes[]))` in calldata
+    /// @param _bytes The input bytes string to extract input arrays from
+    /// @return actions The uint256 calldata array of actions
+    /// @return params The bytes calldata array of parameters
+    function decodeInCalldata(bytes calldata _bytes)
```

Comment:
Hmm.. so I think it makes sense to keep the params in calldata before we call `unlock` but is there benefit to keeping them in calldata here? at some point they are still put in memory.. I'm just wondering if we are doing extra work by doing calldata -> decode but keep in calldata -> memory

and if it would actually be cheaper to just put these directly into memory now?

Reply:
Its for gas efficiency. For some of them they will never need to go into memory - and for the others its cheapest to decode as late as possible so you dont pass around in memory

Reply:
And then when we do decode to memory it should be done in assembly like how UR does it - its muchhh cheaper than doing abi.decode